src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -359,6 +359,10 @@ Klass* SystemDictionary::resolve_array_class_or_null(Symbol* class_name,
                                                          protection_domain,
                                                          CHECK_NULL);
     if (k != NULL) {
+      if ((class_name->is_Q_array_signature() && !k->is_inline_klass()) ||
+          (!class_name->is_Q_array_signature() && k->is_inline_klass())) {
+            THROW_NULL(vmSymbols::java_lang_IncompatibleClassChangeError());
+          }
```

Comment:
Is it worthwhile using THROW_MSG_NULL to provide a helpful message with the exception?

Reply:
Something like this?

`THROW_MSG_NULL(vmSymbols::java_lang_IncompatibleClassChangeError(), "L/Q mismatch on bottom type");`

Fred

Reply:
Yes. That looks good.  Thanks.